File: static/js/components/tradeWindow.js
/* Trading-window editor (entries only run inside a local time range on given
   weekdays). Shared by the webhook drawer and the subscription controls. */
import { h } from "../ui.js";
import { store } from "../store.js";
import { t } from "../i18n.js";

const DAY_LABELS = () => [["mon", t("Mon")], ["tue", t("Tue")], ["wed", t("Wed")], ["thu", t("Thu")], ["fri", t("Fri")], ["sat", t("Sat")], ["sun", t("Sun")]];

/** tradeWindowEditor(initial, { hint }) → { el, collect } */
export function tradeWindowEditor(initial, { hint = null, title = null } = {}) {
  const tw = { enabled: false, from: "08:00", to: "17:00", tz: "", days: ["mon", "tue", "wed", "thu", "fri"], ...(initial || {}) };
  const uid = `tw-${Math.random().toString(36).slice(2, 8)}`;
  const on = h("input", { type: "checkbox", class: "switch", checked: !!tw.enabled });
  const from = h("input", { type: "time", value: tw.from, style: "width:120px" });
  const to = h("input", { type: "time", value: tw.to, style: "width:120px" });
  const tz = h("input", { type: "text", value: tw.tz || "", placeholder: (store.get("settings") || {}).journal_timezone || "Europe/Zurich", style: "width:200px", list: `${uid}-tz` });
  const days = h("div", { class: "check-list", style: "display:flex;flex-direction:row;flex-wrap:wrap;gap:6px 14px;max-height:none" },
    DAY_LABELS().map(([k, label]) => h("label", { class: "check-item" }, h("input", { type: "checkbox", class: "tw-day", value: k, checked: (tw.days || []).includes(k) }), " ", label)));
  const body = h("div", { class: "grid grid-2", style: "margin-top:10px" },
    h("div", { class: "field" }, h("label", null, t("From")), from),
    h("div", { class: "field" }, h("label", null, t("To")), to, h("div", { class: "field-hint" }, t("End before start = spans midnight (22:00 → 06:00)."))),
    h("div", { class: "field" }, h("label", null, t("Weekdays")), days),
    h("div", { class: "field" }, h("label", null, t("Timezone")), tz, h("datalist", { id: `${uid}-tz` }, ["Europe/Zurich", "Europe/London", "America/New_York", "America/Chicago", "UTC"].map((z) => h("option", { value: z }))),
      h("div", { class: "field-hint" }, t("Empty = the journal timezone (Settings → General)."))));
  const sync = () => body.classList.toggle("hidden", !on.checked);
  on.addEventListener("change", sync); sync();
  const el = h("div", { style: "margin-top:16px" },
    h("label", { class: "switch-row" }, h("span", null, title || t("Trading window"), h("small", null, hint || t("Entries (buy / sell, TS-Hunter signals) only run inside this local time range on these weekdays. Closes, stop moves and management signals always run — an open position is never trapped."))), on),
    body);
  const collect = () => ({ enabled: on.checked, from: from.value || "08:00", to: to.value || "17:00", tz: tz.value.trim(),
    days: [...days.querySelectorAll(".tw-day")].filter((c) => c.checked).map((c) => c.value) });
  return { el, collect };
}

File: app/trade_window.py
"""Per-webhook trading window: entries (buy / sell, TS-Hunter signals) are only
executed inside the configured local time range on the configured weekdays.
Exits, stop moves and management signals always run — a window closes the
door for *new* risk, it never traps an open position.

Stored on the webhook as ``trade_window``::

    {"enabled": bool, "from": "08:00", "to": "17:00", "tz": "Europe/Zurich" | "",
     "days": ["mon", "tue", "wed", "thu", "fri"]}

``from`` > ``to`` means the window spans midnight (22:00 → 06:00: the day check
applies to the evening the window opens on). An empty ``tz`` follows the
workspace's journal timezone."""
from __future__ import annotations

from datetime import datetime, timedelta
from typing import Any, Optional
from zoneinfo import ZoneInfo

DAYS = ("mon", "tue", "wed", "thu", "fri", "sat", "sun")
DEFAULT: dict[str, Any] = {"enabled": False, "from": "08:00", "to": "17:00", "tz": "", "days": ["mon", "tue", "wed", "thu", "fri"]}


def _hhmm(raw: Any, default: str) -> str:
    text = str(raw if raw not in (None, "") else default).strip()
    parts = text.split(":")
    if len(parts) != 2 or not all(p.isdigit() for p in parts) or not (0 <= int(parts[0]) < 24 and 0 <= int(parts[1]) < 60):
        raise ValueError(f"time must be HH:MM, got '{text}'")
    return f"{int(parts[0]):02d}:{int(parts[1]):02d}"


def normalize(raw: Any) -> dict[str, Any]:
    """A validated window dict (raises ValueError). ``None`` / ``{}`` → disabled default."""
    if raw in (None, "", False):
        return dict(DEFAULT, days=list(DEFAULT["days"]))
    if not isinstance(raw, dict):
        raise ValueError("trade_window must be an object")
    out = {"enabled": bool(raw.get("enabled")), "from": _hhmm(raw.get("from"), DEFAULT["from"]), "to": _hhmm(raw.get("to"), DEFAULT["to"])}
    tz = str(raw.get("tz") or "").strip()
    if tz:
        try:
            ZoneInfo(tz)
        except Exception as exc:  # noqa: BLE001
            raise ValueError(f"unknown timezone '{tz}' (use an IANA name like Europe/Zurich)") from exc
    out["tz"] = tz[:64]
    days_raw = raw.get("days")
    if days_raw is None:
        days = list(DEFAULT["days"])
    elif isinstance(days_raw, (list, tuple)):
        days = [str(d).lower()[:3] for d in days_raw]
        bad = [d for d in days if d not in DAYS]
        if bad:
            raise ValueError(f"unknown weekday '{bad[0]}' (mon … sun)")
        days = [d for d in DAYS if d in days]
    else:
        raise ValueError("days must be a list of weekdays")
    if out["enabled"] and not days:
        raise ValueError("an enabled trading window needs at least one weekday")
    if out["enabled"] and out["from"] == out["to"]:
        raise ValueError("the window's start and end must differ")
    out["days"] = days
    return out


def _zone(name: str, fallback: str) -> ZoneInfo:
    for cand in (name, fallback, "Europe/Zurich"):
        if cand:
            try:
                return ZoneInfo(cand)
            except Exception:  # noqa: BLE001
                continue
    return ZoneInfo("UTC")


def is_open(window: Any, *, now: Optional[datetime] = None, default_tz: str = "") -> tuple[bool, str]:
    """``(True, "")`` when entries may run now, else ``(False, reason)``. A
    missing or disabled window is always open; a malformed one is treated as
    open too (a broken setting must not silently stop a strategy) — the API
    never stores a malformed one."""
    if not isinstance(window, dict) or not window.get("enabled"):
        return True, ""
    try:
        w = normalize(window)
    except ValueError:
        return True, ""
    zone = _zone(w["tz"], default_tz)
    local = (now or datetime.now(zone)).astimezone(zone)
    hm = local.strftime("%H:%M")
    day = DAYS[local.weekday()]
    frm, to, days = w["from"], w["to"], w["days"]
    if frm < to:
        opened = day in days and frm <= hm < to
    else:                                              # spans midnight: 22:00 → 06:00
        prev = DAYS[(local - timedelta(days=1)).weekday()]
        opened = (day in days and hm >= frm) or (prev in days and hm < to)
    if opened:
        return True, ""
    label = f"{frm}–{to} {zone.key}" if hasattr(zone, "key") else f"{frm}–{to}"
    return False, f"outside the trading window {label} ({', '.join(days)}); now {hm} {day}"
